### PR TITLE
[libc++] Deprecate the C++20 synchronization library before C++20

### PR DESCRIPTION
**Pull request by ldionne**
When we initially implemented the C++20 synchronization library, we reluctantly accepted for the implementation to be backported to C++03 upon request from the person who provided the patch. This was when we were only starting to have experience with the issues this can create, so we flinched. Nowadays, we have a much stricter stance about not backporting features to previous standards.

We have recently started fixing several bugs (and near bugs) in our implementation of the synchronization library. A recurring theme during these reviews has been how difficult to understand the current code is, and upon inspection it becomes clear that being able to use a few recent C++ features (in particular lambdas) would help a great deal. The code would still be pretty intricate, but it would be a lot easier to reason about the flow of callbacks through things like __thread_poll_with_backoff.

As a result, this patch deprecates support for the synchronization library before C++20. In the next release, we can remove that support entirely.

### CODE REVIEW

**Comment by ldionne** [comment]
This is the deprecation step before #82008.

Note that this PR is incomplete, I'll have another pass at it.

File: libcxx/docs/ReleaseNotes/19.rst [hunk 1]
```diff
@@ -55,6 +55,10 @@ Improvements and New Features
 Deprecations and Removals
 -------------------------
 
+- The C++20 synchronization library (``<barrier>``, ``<latch>``, ``atomic::wait``, etc.) has been deprecated
+  in language modes prior to C++20. If you are using these features prior to C++20, please update to ``-std=c++20``.
+  In LLVM 20, the C++20 synchronization library will be removed entirely in language modes prior to C++20.
```
Comment (mordante) [inline libcxx/docs/ReleaseNotes/19.rst:60]: Don't we typically deprecate for 2 releases?

File: libcxx/include/__config [hunk 2]
```diff
@@ -954,6 +954,14 @@ typedef __char32_t char32_t;
 #    define _LIBCPP_DEPRECATED_(m)
 #  endif
 
+#  if _LIBCPP_STD_VER < 20
+#    define _LIBCPP_DEPRECATED_ATOMIC_SYNC                                                                             \
+      _LIBCPP_DEPRECATED_("The C++20 synchronization library has been deprecated prior to C++20. Please update to "    \
+                          "using -std=c++20 if you need to use these facilities")
```
Comment (mordante) [inline libcxx/include/__config:960]:
```suggestion
                          "using -std=c++20 if you need to use these facilities.")
```

**Review by mordante — CHANGES_REQUESTED** [review]
Thanks for the patch! As expected the CI fails; Some tests need to allow deprecated messages. Maybe add a TODO there so we remove it once we remove the deprecated feature.

**Comment by EricWF** [comment]
I would argue this is a unfortunate use for the deprecation warning, given that in the future the thing is not deprecated, we're just deprecating backwards support for it?

I would also be careful with this because synchronization is one of those things that we may want to use internally prior to C++20.

**Comment by EricWF** [comment]
I actually think this might be a case where providing the extension costs us less than removing it.

**Comment by ldionne** [comment]
> I would argue this is a unfortunate use for the deprecation warning, given that in the future the thing is not deprecated, we're just deprecating backwards support for it?

Yes, but the deprecation warning includes a message explaining that it's only the use of it before C++20 that is deprecated, so I think this is not ambiguous.

> 
> I would also be careful with this because synchronization is one of those things that we may want to use internally prior to C++20.

In that case, we'd need to create an internal-only API like we often do in other cases. I don't see a problem with that *if* the need arises. I also think we're unlikely to want to use it for pre-C++20 stuff at this point since we implement almost everything pre-C++20 and we haven't had a need for it so far.

> I actually think this might be a case where providing the extension costs us less than removing it.

I have to disagree. The synchronization library is currently really complex and has (or at least had) several subtle bugs. Some of those can be tracked down in part to the difficulty to reason about stuff like `__libcpp_thread_poll_with_backoff`'s flow which would be a lot clearer if we localized e.g. `__atomic_wait_poll_impl`, but that requires lambdas. This desire to deprecate and remove didn't come out of nowhere, it came out of hours of @huixie90 and I plowing through the implementation trying to understand it and often thinking "wow this would have been a lot easier to understand if we had this or that feature".

File: libcxx/docs/ReleaseNotes/19.rst (hunk 1 above)
Comment (ldionne) [reply libcxx/docs/ReleaseNotes/19.rst:60]: I don't think we do? Or at least I don't think we have any sort of fixed policy for doing that?

I'm pretty sure we've deprecated stuff in one release and removed it in the next in some cases, and we might also have kept stuff deprecated for 2 release before removing it. Or is my memory playing tricks on me and we've actually always deprecated for 2 releases before removing? CC @philnik777

**Comment by ldionne** [comment]
Pinging @llvm/libcxx-vendors for awareness.

File: libcxx/test/std/atomics/atomics.types.operations/atomics.types.operations.wait/atomic_notify_all.pass.cpp [hunk 3]
```diff
@@ -7,7 +7,7 @@
 //===----------------------------------------------------------------------===//
 //
 // UNSUPPORTED: no-threads
-// XFAIL: c++03
+// UNSUPPORTED: c++03, c++11, c++14, c++17
```
Comment (mordante) [inline libcxx/test/std/atomics/atomics.types.operations/atomics.types.operations.wait/atomic_notify_all.pass.cpp:10]: Normally we use `// ADDITIONAL_COMPILE_FLAGS: -D_LIBCPP_DISABLE_DEPRECATION_WARNINGS` to keep testing deprecated features. Should we do the same here?

File: libcxx/docs/ReleaseNotes/19.rst (hunk 1 above)
Comment (mordante) [reply libcxx/docs/ReleaseNotes/19.rst:60]: I checked and it indeed seems we use a shorter deprecation (nowadays?). So I'm oke to keep this as is.

**Review by mordante — APPROVED** [review]
Basically LGTM but I prefer to keep the tests working until we remove support.

**Comment by EricWF** [comment]
> I have to disagree. The synchronization library is currently really complex and has (or at least had) several subtle bugs. Some of those can be tracked down in part to the difficulty to reason about stuff like `__libcpp_thread_poll_with_backoff`'s flow which would be a lot clearer if we localized e.g. `__atomic_wait_poll_impl`, but that requires lambdas. This desire to deprecate and remove didn't come out of nowhere, it came out of hours of @huixie90 and I plowing through the implementation trying to understand it and often thinking "wow this would have been a lot easier to understand if we had this or that feature".

I'll take your word on that. The surface level complexity of the functions we've applied the deprecation macro to don't immediately make that clear.

**Review by EricWF — APPROVED** [review]
I have no issue with this. It looks good to me, but I defer to the other reviewers.

**Comment by ldionne** [comment]
> I'll take your word on that. The surface level complexity of the functions we've applied the deprecation macro to don't immediately make that clear.

That's right. The surface functions are really simple but the implementation is where it gets more complicated.

**Pull request merged**